janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/optimize/step/JanusGraphVertexStep.java

```diff
@@ -108,10 +108,15 @@ public void setTxVertexCacheSize(int txVertexCacheSize) {
      * This initialisation method is called when an attempt to retrieve a vertex from the cached multiQuery results
      * doesn't find an entry.
      */
-    private void prefetchNextBatch() {
+    private void prefetchNextBatch(final Traverser.Admin<Vertex> traverser) {
         final JanusGraphMultiVertexQuery multiQuery = JanusGraphTraversalUtil.getTx(getTraversal()).multiQuery();
-        multiQuery.addAllVertices(verticesToPrefetch);
-        verticesToPrefetch.clear();
+        if(verticesToPrefetch.isEmpty() && traverser!=null){
+            multiQuery.addVertex(traverser.get());
```

Comment:
Should we use `JanusGraphTraversalUtil.getJanusGraphVertex(traverser)` here?

Reply:
I was also thinking about that but it's not Traversal but Traveser, thus types are incomplete

Reply:
```suggestion
            multiQuery.addVertex(JanusGraphTraversalUtil.getJanusGraphVertex(traverser));
```
this works for me.

Reply:
Ah, you are right. I confused with another method.

I also replaced `if(verticesToPrefetch.isEmpty() && traverser!=null)` with `if(verticesToPrefetch.isEmpty())` because `traverser` should not be `null` here. Otherwise we will throw NPE just after this method.